### PR TITLE
MVPoly: add is_multilinear

### PR DESCRIPTION
**Pull request by dannywillems**

### CODE REVIEW

**Review by marcbeunardeau88 — APPROVED** [review]

File: mvpoly/src/prime.rs [hunk 1]
```diff
@@ -448,6 +448,26 @@ impl<F: PrimeField, const N: usize, const D: usize> MVPoly<F, N, D> for Dense<F,
     ) -> HashMap<usize, F> {
         unimplemented!()
     }
+
+    fn is_multilinear(&self) -> bool {
+        if self.is_zero() {
+            return true;
+        }
```
Comment (Fizzixnerd) [inline mvpoly/src/prime.rs:455]: Piggybacking on what Marc said, it seems you consider a zero polynomial to be multilinear here.

File: mvpoly/src/monomials.rs [hunk 2]
```diff
@@ -564,6 +564,12 @@ impl<const N: usize, const D: usize, F: PrimeField> MVPoly<F, N, D> for Sparse<F
         });
         cross_terms_by_powers_of_r
     }
+
+    fn is_multilinear(&self) -> bool {
+        self.monomials
+            .iter()
+            .all(|(exponents, _)| exponents.iter().all(|&d| d <= 1))
```
[the remark this answers is not in the document]
Comment (Fizzixnerd) [reply mvpoly/src/monomials.rs:571]: In addition, this is incorrect, as if d == 0, then it's not linear.  For example, X * Y + 1 is affine, not linear.

File: mvpoly/tests/monomials.rs [hunk 3]
```diff
@@ -663,3 +663,38 @@ fn test_mvpoly_compute_cross_terms_degree_seven() {
...
+        let mut monomials_exponents: [usize; 6] =
+            std::array::from_fn(|i| if i <= nb_var { 1 } else { 0 });
+        monomials_exponents.shuffle(&mut rng);
+        p.add_monomial(monomials_exponents, c);
+        assert!(p.is_multilinear());
+    }
+
+    // Very unlikely to have a random polynomial being multilinear
+    let p3 = unsafe { Sparse::<Fp, 6, 4>::random(&mut rng, None) };
+    assert!(!p3.is_multilinear());
```
Comment (Fizzixnerd) [inline mvpoly/tests/monomials.rs:699]: Is there a way to mark this test as flakey?  Maybe I just don't understand _how_ unlikely it is, if it's like 1 / |F| then I guess it's fine.

File: mvpoly/src/monomials.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dannywillems) [reply mvpoly/src/monomials.rs:571]: >  It's unclear to me if not having zeroes coefficients is an invariant of the lib

Yes, it is an invariant.

File: mvpoly/src/prime.rs (hunk 1 above)
Comment (dannywillems) [reply mvpoly/src/prime.rs:455]: It is.

File: mvpoly/src/monomials.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fizzixnerd) [reply mvpoly/src/monomials.rs:571]: After speaking with Danny, it seems we are using different definitions of the term 'multilinear'.  I'm good to resolve.

**Review by Fizzixnerd — APPROVED** [review]

**Pull request merged**